### PR TITLE
GH-41749: [GLib] Allow getting a RecordBatchReader from a Dataset or Scanner

### PR DESCRIPTION
**Pull request by adamreeve**
### Rationale for this change

See #41749

### What changes are included in this PR?

Adds `to_reader` methods to `GADatasetDataset` and `GADatasetScanner`.

### Are these changes tested?

Yes I've added new unit tests.

### Are there any user-facing changes?

Yes this is a new feature.
* GitHub Issue: #41749

### CODE REVIEW

File: c_glib/arrow-glib/version.h.in [hunk 1]
```diff
@@ -108,6 +108,15 @@
...
 #endif
 
+/**
+ * GARROW_VERSION_17_0:
+ *
+ * You can use this macro value for compile time API version check.
+ *
+ * Since: 17.0.0
+ */
+#define GARROW_VERSION_17_0 G_ENCODE_VERSION(17, 0)
```
Comment (adamreeve) [inline c_glib/arrow-glib/version.h.in:118]: This will conflict with #41721

File: c_glib/arrow-dataset-glib/dataset.cpp [hunk 2]
```diff
@@ -152,12 +153,43 @@ gadataset_dataset_to_table(GADatasetDataset *dataset, GError **error)
   }
   auto arrow_scanner = *arrow_scanner_result;
   auto arrow_table_result = arrow_scanner->ToTable();
-  if (!garrow::check(error, arrow_scanner_result, "[dataset][to-table]")) {
+  if (!garrow::check(error, arrow_table_result, "[dataset][to-table]")) {
```
Comment (adamreeve) [inline c_glib/arrow-dataset-glib/dataset.cpp:156]: This is an unrelated change but I noticed this was checking the wrong result, so if `arrow_scanner_builder->Finish()` succeeded but `arrow_scanner->ToTable()` failed the error would not have been correctly caught.

File: c_glib/test/dataset/test-file-system-dataset.rb [hunk 3]
```diff
@@ -56,6 +56,24 @@ def test_partitioning
...
+    dataset, expected_table = create_dataset
+    reader = dataset.to_reader
+    record_batches = []
+    loop do
+      batch = reader.read_next
+      break if batch.nil?
+      record_batches << batch
+    end
+    table_from_batches = Arrow::Table.new(reader.schema, record_batches)
+    assert_equal(expected_table, table_from_batches)
```
Comment (kou) [inline c_glib/test/dataset/test-file-system-dataset.rb:73]:
```suggestion
    assert_equal(expected_table, reader.read_all)
```

File: c_glib/test/dataset/test-scanner.rb [hunk 4]
```diff
@@ -45,4 +45,16 @@ def setup
...
+  def test_record_batch_reader
+    reader = @scanner.to_reader
+    record_batches = []
+    loop do
+      batch = reader.read_next
+      break if batch.nil?
+      record_batches << batch
+    end
+    table_from_batches = Arrow::Table.new(reader.schema, record_batches)
+    assert_equal(@table, table_from_batches)
```
Comment (kou) [inline c_glib/test/dataset/test-scanner.rb:58]:
```suggestion
    assert_equal(@table, @scanner.to_reader.read_all)
```

File: c_glib/arrow-dataset-glib/dataset.cpp [hunk 5]
```diff
@@ -152,12 +153,43 @@ gadataset_dataset_to_table(GADatasetDataset *dataset, GError **error)
...
+ *   A #GArrowRecordBatchReader on success, %NULL on error.
+ *
+ * Since: 17.0.0
+ */
+GArrowRecordBatchReader *
+gadataset_dataset_to_reader(GADatasetDataset *dataset, GError **error)
+{
+  auto arrow_dataset = gadataset_dataset_get_raw(dataset);
+  auto arrow_scanner_builder_result = arrow_dataset->NewScan();
+  if (!garrow::check(error, arrow_scanner_builder_result, "[dataset][get-reader]")) {
```
Comment (kou) [inline c_glib/arrow-dataset-glib/dataset.cpp:177]:
```suggestion
  if (!garrow::check(error, arrow_scanner_builder_result, "[dataset][to-reader]")) {
```

File: c_glib/arrow-dataset-glib/dataset.cpp [hunk 6]
```diff
@@ -152,12 +153,43 @@ gadataset_dataset_to_table(GADatasetDataset *dataset, GError **error)
...
+    return NULL;
+  }
+  auto arrow_scanner_builder = *arrow_scanner_builder_result;
+  auto arrow_scanner_result = arrow_scanner_builder->Finish();
+  if (!garrow::check(error, arrow_scanner_result, "[dataset][get-reader]")) {
```
Comment (kou) [inline c_glib/arrow-dataset-glib/dataset.cpp:182]:
```suggestion
  if (!garrow::check(error, arrow_scanner_result, "[dataset][to-reader]")) {
```

File: c_glib/arrow-dataset-glib/dataset.cpp [hunk 7]
```diff
@@ -152,12 +153,43 @@ gadataset_dataset_to_table(GADatasetDataset *dataset, GError **error)
...
+    return NULL;
+  }
+  auto arrow_scanner = *arrow_scanner_result;
+  auto arrow_reader_result = arrow_scanner->ToRecordBatchReader();
+  if (!garrow::check(error, arrow_reader_result, "[dataset][get-reader]")) {
```
Comment (kou) [inline c_glib/arrow-dataset-glib/dataset.cpp:187]:
```suggestion
  if (!garrow::check(error, arrow_reader_result, "[dataset][to-reader]")) {
```

File: c_glib/arrow-dataset-glib/scanner.cpp [hunk 8]
```diff
@@ -128,6 +128,27 @@ gadataset_scanner_to_table(GADatasetScanner *scanner, GError **error)
...
+ *   A #GArrowRecordBatchReader on success, %NULL on error.
+ *
+ * Since: 17.0.0
+ */
+GArrowRecordBatchReader *
+gadataset_scanner_to_reader(GADatasetScanner *scanner, GError **error)
+{
+  auto arrow_scanner = gadataset_scanner_get_raw(scanner);
+  auto arrow_reader_result = arrow_scanner->ToRecordBatchReader();
+  if (!garrow::check(error, arrow_reader_result, "[scanner][get-reader]")) {
```
Comment (kou) [inline c_glib/arrow-dataset-glib/scanner.cpp:146]:
```suggestion
  if (!garrow::check(error, arrow_reader_result, "[scanner][to-reader]")) {
```

File: c_glib/arrow-dataset-glib/dataset.cpp [hunk 9]
```diff
@@ -152,12 +153,43 @@ gadataset_dataset_to_table(GADatasetDataset *dataset, GError **error)
...
+ * @dataset: A #GADatasetDataset.
+ * @error: (nullable): Return location for a #GError or %NULL.
+ *
+ * Returns: (transfer full) (nullable):
+ *   A #GArrowRecordBatchReader on success, %NULL on error.
+ *
+ * Since: 17.0.0
+ */
+GArrowRecordBatchReader *
+gadataset_dataset_to_reader(GADatasetDataset *dataset, GError **error)
```
Comment (kou) [inline c_glib/arrow-dataset-glib/dataset.cpp:173]: How about using `to_record_batch_reader` not just `to_reader`? We have many readers in our API.

Comment (adamreeve) [reply c_glib/arrow-dataset-glib/dataset.cpp:173]: Yeah good point, I'll change this.

File: c_glib/test/dataset/test-file-system-dataset.rb (hunk 3 above)
Comment (adamreeve) [reply c_glib/test/dataset/test-file-system-dataset.rb:73]: That's much simpler thanks, not sure how I missed that :smile:

**Comment by adamreeve** [comment]
I'm not sure how I've broken the tests on MinGW, I'll need to figure out what's going wrong there.

**Comment by kou** [comment]
Could you try this?

```diff
diff --git a/c_glib/test/dataset/test-file-system-dataset.rb b/c_glib/test/dataset/test-file-system-dataset.rb
index a64022b618..d4bf567397 100644
--- a/c_glib/test/dataset/test-file-system-dataset.rb
+++ b/c_glib/test/dataset/test-file-system-dataset.rb
@@ -63,7 +63,11 @@ class TestDatasetFileSystemDataset < Test::Unit::TestCase
   def test_record_batch_reader
     dataset, expected_table = create_dataset
     reader = dataset.to_record_batch_reader
-    assert_equal(expected_table, reader.read_all)
+    begin
+      assert_equal(expected_table, reader.read_all)
+    ensure
+      reader.close
+    end
   end
 
   def create_dataset
diff --git a/c_glib/test/dataset/test-scanner.rb b/c_glib/test/dataset/test-scanner.rb
index 714760d6a3..1a50b7fe68 100644
--- a/c_glib/test/dataset/test-scanner.rb
+++ b/c_glib/test/dataset/test-scanner.rb
@@ -47,6 +47,11 @@ class TestDatasetScanner < Test::Unit::TestCase
   end
 
   def test_record_batch_reader
-    assert_equal(@table, @scanner.to_record_batch_reader.read_all)
+    reader = @scanner.to_record_batch_reader
+    begin
+      assert_equal(@table, reader.read_all)
+    ensure
+      reader.close
+    end
   end
 end
```

**Comment by adamreeve** [comment]
> Could you try this?

That doesn't work because there aren't bindings for the `Close` method, but even if there were, it looks like this isn't sufficient to close the underlying files. I think this is a bug in the C++ implementation, I've opened #41771 for this.

I guess for now I could manually cleanup the temporary directory and use the `force` option ignore errors.

File: c_glib/test/dataset/test-scanner.rb [hunk 10]
```diff
@@ -39,10 +40,22 @@ def setup
...
+      GC.start
+      FileUtils.remove_entry(tmpdir, force: true)
     end
   end
 
   def test_to_table
     assert_equal(@table, @scanner.to_table)
   end
+
+  def test_record_batch_reader
```
Comment (kou) [inline c_glib/test/dataset/test-scanner.rb:58]:
```suggestion
  def test_to_record_batch_reader
```

File: c_glib/test/dataset/test-file-system-dataset.rb [hunk 11]
```diff
@@ -56,6 +65,17 @@ def test_partitioning
   end
 
   def test_read_write
+    dataset, expected_table = create_dataset
+    assert_equal(expected_table, dataset.to_table)
+  end
+
+  def test_record_batch_reader
```
Comment (kou) [inline c_glib/test/dataset/test-file-system-dataset.rb:72]:
```suggestion
  def test_to_record_batch_reader
```

**Review by kou — COMMENTED** [review]
Does this work?

```diff
diff --git a/c_glib/test/dataset/test-file-system-dataset.rb b/c_glib/test/dataset/test-file-system-dataset.rb
index ac67c6d61d..972c2d2ef9 100644
--- a/c_glib/test/dataset/test-file-system-dataset.rb
+++ b/c_glib/test/dataset/test-file-system-dataset.rb
@@ -21,8 +21,7 @@ class TestDatasetFileSystemDataset < Test::Unit::TestCase
 
   def setup
     omit("Arrow Dataset is required") unless defined?(ArrowDataset)
-    tmpdir = Dir.mktmpdir
-    begin
+    Dir.mktmpdir do |tmpdir|
       @dir = tmpdir
       @format = ArrowDataset::IPCFileFormat.new
       @factory = ArrowDataset::FileSystemDatasetFactory.new(@format)
@@ -33,14 +32,6 @@ class TestDatasetFileSystemDataset < Test::Unit::TestCase
         ArrowDataset::DirectoryPartitioning.new(partitioning_schema)
       @factory.partitioning = @partitioning
       yield
-    ensure
-      # We have to ignore errors trying to remove the directory due to
-      # the RecordBatchReader not closing files
-      # (https://github.com/apache/arrow/issues/41771).
-      # Also request GC first which should free any remaining RecordBatchReader
-      # and close open files.
-      GC.start
-      FileUtils.remove_entry(tmpdir, force: true)
     end
   end
 
@@ -72,7 +63,11 @@ class TestDatasetFileSystemDataset < Test::Unit::TestCase
   def test_record_batch_reader
     dataset, expected_table = create_dataset
     reader = dataset.to_record_batch_reader
-    assert_equal(expected_table, reader.read_all)
+    begin
+      assert_equal(expected_table, reader.read_all)
+    ensure
+      reader.unref
+    end
   end
 
   def create_dataset
diff --git a/c_glib/test/dataset/test-scanner.rb b/c_glib/test/dataset/test-scanner.rb
index ef834674cb..fe3478a866 100644
--- a/c_glib/test/dataset/test-scanner.rb
+++ b/c_glib/test/dataset/test-scanner.rb
@@ -21,8 +21,7 @@ class TestDatasetScanner < Test::Unit::TestCase
 
   def setup
     omit("Arrow Dataset is required") unless defined?(ArrowDataset)
-    tmpdir = Dir.mktmpdir
-    begin
+    Dir.mktmpdir do |tmpdir|
       path = File.join(tmpdir, "table.arrow")
       @table = build_table(visible: [
                              build_boolean_array([true, false, true]),
@@ -40,14 +39,6 @@ class TestDatasetScanner < Test::Unit::TestCase
       builder = @dataset.begin_scan
       @scanner = builder.finish
       yield
-    ensure
-      # We have to ignore errors trying to remove the directory due to
-      # the RecordBatchReader not closing files
-      # (https://github.com/apache/arrow/issues/41771).
-      # Also request GC first which should free any remaining RecordBatchReader
-      # and close open files.
-      GC.start
-      FileUtils.remove_entry(tmpdir, force: true)
     end
   end
 
@@ -56,6 +47,11 @@ class TestDatasetScanner < Test::Unit::TestCase
   end
 
   def test_record_batch_reader
-    assert_equal(@table, @scanner.to_record_batch_reader.read_all)
+    reader = @scanner.to_record_batch_reader
+    begin
+      assert_equal(@table, reader.read_all)
+    ensure
+      reader.unref
+    end
   end
 end
```

**Comment by adamreeve** [comment]
> Does this work?

Yes that works thanks, that's much better. Although it looks like the unref call may be unnecessary after #41824, so should I wait for that to be merged then remove the unref call?

**Review by kou — APPROVED** [review]
+1

We can use `unref` for now because #41824 may be rejected.

**Pull request merged**